plasmapy/diagnostics/proton_radiography.py

```diff
@@ -0,0 +1,1022 @@
+            [[hmin,hmax], [vmin, vmax]]. Units must be convertable to meters.
+
+        bins : array of integers, shape (2)
+            The number of bins in each direction in the format [hbins, vbins].
+            The default is [200,200].
+
+        ignore_grid: bool
+            If True, returns the intensity in the image plane in the absence
+            of simulated fields.
+
+        optical_density: bool
+            If True, return the optical density rather than the intensity
+
+            .. math::
+                OD = -log_{10}(Intensity/I_0)
+
+            where I_O is the intensity on the detector plane in the absence of
+            simulated fields. Default is False.
+
+        Returns
+        -------
+        hax : `~astropy.units.Quantity` array shape (hbins,)
+            The horizontal axis of the synthetic radiograph in meters.
+
+        vax : `~astropy.units.Quantity` array shape (vbins, )
+            The vertical axis of the synthetic radiograph in meters.
+
+        intensity : ndarray, shape (hbins, vbins)
+            The number of protons counted in each bin of the histogram.
+        """
+
+        # Note that, at the end of the simulation, all particles were moved
+        # into the image plane.
+
+        # If ignore_grid is True, use the predicted positions in the absence of
+        # simulated fields
+        if ignore_grid:
+            x = self.x0
+        else:
+            x = self.x
+
+        # Determine locations of points in the detector plane using unit
+        # vectors
+        xloc = np.dot(x - self.detector, self.det_hdir)
+        yloc = np.dot(x - self.detector, self.det_vdir)
+
+        if size is None:
+            # If a detector size is not given, choose lengths based on the
+            # dimensions of the grid
+            w = self.mag * np.max(
+                [
+                    np.max(np.abs(self.grid.pts0.to(u.m).value)),
+                    np.max(np.abs(self.grid.pts1.to(u.m).value)),
+                    np.max(np.abs(self.grid.pts2.to(u.m).value)),
+                ]
+            )
+
+            size = np.array([[-w, w], [-w, w]]) * self.grid.unit
+
+        # Generate the histogram
+        intensity, h, v = np.histogram2d(
+            xloc, yloc, range=size.to(u.m).value, bins=bins
+        )
+
+        # h, v are the bin edges: compute the centers to produce arrays
+        # of the right length (then trim off the extra point)
+        h = ((h + np.roll(h, -1)) / 2)[0:-1]
+        v = ((v + np.roll(v, -1)) / 2)[0:-1]
+
+        # Throw a warning if < 50% of the particles are included on the
+        # histogram
+        percentage = 100 * np.sum(intensity) / self.nparticles
+        if percentage < 50:
+            warnings.warn(
+                f"Only {percentage:.2f}% of the particles are shown "
+                " on this synthetic radiograph. Consider increasing "
+                " the size to include more.",
+                RuntimeWarning,
+            )
```

Comment:
This is one of my new favorite tricks:

```suggestion
        # Throw a warning if < 50% of the particles are included on the
        # histogram
        percentage = np.sum(intensity) / self.nparticles
        if percentage < 0.5:
            warnings.warn(
                f"Only {percentage:.2%} of the particles are shown "
                " on this synthetic radiograph. Consider increasing "
                " the size to include more.",
                RuntimeWarning,
            )
```

Reply:
I didn't know that was an option!

Reply:
I learned it a few days ago! :D